README.md
# codice ficale

[![Latest Version on Packagist](https://img.shields.io/packagist/v/ivanomatteo/codice-fiscale.svg?style=flat-square)](https://packagist.org/packages/ivanomatteo/codice-fiscale)
[![Total Downloads](https://img.shields.io/packagist/dt/ivanomatteo/codice-fiscale.svg?style=flat-square)](https://packagist.org/packages/ivanomatteo/codice-fiscale)

Calculation and data extraction on italian fiscal code (codice fiscale), with "omocodie". 


It also support diacritical character expansion (caratteri diacritici)
for example:

'Ä' => 'AE',  'ß' => 'SS'

...


## Installation

You can install the package via composer:

```bash
composer require ivanomatteo/codice-fiscale
```

## Usage

``` php
include __DIR__.'/vendor/autoload.php';

use IvanoMatteo\CodiceFiscale\CodiceFiscale;


try{

    // will raise an exception if the format is not valid
    $c = CodiceFiscale::parse("RSSMRAULRL1H50MM",1900); // passing the "century" arg, help to validate in case of leap years
    $c = CodiceFiscale::parse("RSSMRAULRL1H50MM");

    echo "\n"."matchName ".($c->matchName('Mario')? 'si' : 'no');
    echo "\n"."isOmocodia ".($c->isOmocodia()? 'si' : 'no');

    // to extract the date of birth, the "century" argumenti is required
    echo "\n"."getDateOfBirth ".$c->getDateOfBirth(1900)->format('Y-m-d');

    // is possible to extract the raw date of birth:  { "year":"yy", "month":"mm", "day":"dd" }
    echo "\n"."getDateOfBirthObj ".json_encode($c->getDateOfBirthRaw());

    // the library can provide the nearest date of birth according to current date:
    echo "\n"."getProbableDateOfBirth ".$c->getProbableDateOfBirth()->format('Y-m-d');
    // also according to minimum age and current date arguments
    echo "\n"."getProbableDateOfBirth 18 ".$c->getProbableDateOfBirth(18,'2019-01-01')->format('Y-m-d');

}catch(\Exception $ex){
  echo $ex->getMessage();
}


//
// calculate using person data 
//
$name = 'Mario';
$familyName = 'Rossi';
$dateOfBirth = '1980-10-01';
$sex = 'M';
$cityCode = 'H501';

$cf = CodiceFiscale::calculate($name, $familyName, $dateOfBirth, $sex, $cityCode);

//
// calcumate using an array (it accept an object as well)
//
$person = [
    'name' => $name,
    'familyName' => $familyName,
    'date' => $dateOfBirth,
    'sex' => $sex,
    'cityCode' => $cityCode,
];
$map = [  // it possible to provide a map, to match fields with different names
    'dateOfBirth' => 'date'
];
$cfx = CodiceFiscale::calculateObj($person, $map);


// generate all the 127 possible "omocodia" variations
$variazioni = $cf->generateVariations();

// generate variation n. 7 only
$variazioni = $cf->generateVariations(7);


```

### Testing

``` bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email ivanomatteo@gmail.com instead of using the issue tracker.

## Credits

- [Ivano Matteo](https://github.com/ivanomatteo)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.



CHANGELOG.md
# Changelog

All notable changes to `codicefiscale` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/CodFiscTest.php
<?php

namespace IvanoMatteo\CodiceFiscale\Tests;

use IvanoMatteo\CodiceFiscale\CodiceFiscale;
use IvanoMatteo\CodiceFiscale\CodicefiscaleException;
use Orchestra\Testbench\TestCase;


class CodFiscTest extends TestCase
{

    /** @test */
    public function test_vari()
    {
        echo "\n";

        $name = 'Mario';
        $familyName = 'Rossi';
        $dateOfBirth = '1980-10-01';
        $sex = 'M';
        $cityCode = 'H501';

        $cf = CodiceFiscale::calculate($name, $familyName, $dateOfBirth, $sex, $cityCode);



        $person = [
            'name' => $name,
            'familyName' => $familyName,
            'date' => $dateOfBirth,
            'sex' => $sex,
            'cityCode' => $cityCode,
        ];
        $fieldMap = [
            'dateOfBirth' => 'date'
        ];

        $cfx = CodiceFiscale::calculateObj($person, $fieldMap);

        $this->assertTrue($cf->__toString() === $cfx->__toString());

        echo "Cofice Fiscale: $cf\n\n";

        echo "variazione[7]: " . $cf->generateVariations(7) . "\n";

        $variazioni = $cf->generateVariations();
        print_r($variazioni);

        $variazioni[] = ('' . $cf);

        foreach ($variazioni as $cod) {
            $c = CodiceFiscale::parse($cod);

            $this->assertTrue($c->matchName($name));
            $this->assertTrue($c->matchFamilyName($familyName));
            $this->assertTrue($c->matchCityCode($cityCode));
            $this->assertTrue($c->matchDateOfBirth($dateOfBirth));
            $this->assertTrue($c->matchSex($sex));

            $this->assertTrue(null === $c->validate((object) compact('name', 'familyName', 'dateOfBirth', 'cityCode', 'sex')));
            $this->assertTrue(null === $c->validate($person,$fieldMap));

            //partial validation
            $this->assertTrue(null === $c->validate(compact('name',  'dateOfBirth', 'cityCode'),null,true));
            $this->assertTrue(null !== $c->validate(['name' => 'Mario', 'cityCode'=> 'Z000'],null,true));

            $this->assertTrue($c->getCityCode() === $cityCode);
            $this->assertTrue($c->getSex() === $sex);
            $this->assertTrue($c->getDateOfBirth(1900)->format('Y-m-d') === $dateOfBirth);
        }


        $this->assertTrue($c->getDateOfBirth(1900)->format('Y-m-d') === $dateOfBirth);


        $cfStr = 'RSSMRA81B29H501B';
        $this->assertTrue(CodiceFiscale::strMatchFormat($cfStr));
        $this->assertFalse(CodiceFiscale::isDateOfBirthCorrect($cfStr));
        $this->assertFalse(CodiceFiscale::strMatchControlDigit($cfStr));



        $this->assertTrue(CodiceFiscale::calculateProbableDateOfBirth('12', '01', '01', null, mktime(0, 0, 0, 1, 1, 2020))->format('Y-m-d') === '2012-01-01');
        $this->assertTrue(CodiceFiscale::calculateProbableDateOfBirth('12', '01', '01', 18, mktime(0, 0, 0, 1, 1, 2020))->format('Y-m-d') === '1912-01-01');

        $this->assertTrue(CodiceFiscale::calculateProbableDateOfBirth('20', '01', '01', null, mktime(0, 0, 0, 2, 1, 2020))->format('Y-m-d') === '2020-01-01');
        $this->assertTrue(CodiceFiscale::calculateProbableDateOfBirth('20', '01', '01', null, mktime(0, 0, 0, 1, 1, 2020))->format('Y-m-d') === '1920-01-01');
    }

    /** @test */
    public function test_db()
    {
        $envfile = __DIR__ . '/../env.php';
        if (!file_exists($envfile)) {
            $this->assertTrue(true);
            return;
        }

        $cfg = include($envfile);
        $cfg = $cfg['db'];

        $conn = new \PDO('mysql:host=' . $cfg['host'] . ';dbname=' . $cfg['db'], $cfg['user'], $cfg['pass']);
        $query = $conn->query("SELECT codiceFiscale FROM " . $cfg['table']);

        foreach ($query as $row) {

            if (!empty($row['codiceFiscale'])) {
                try {
                    $c = CodiceFiscale::parse($row['codiceFiscale']);

                    if ($c->isOmocodia()) {
                        echo "omocodia: " . $row['codiceFiscale'] . "\n";
                    }

                    $this->assertTrue((bool) preg_match("/^[A-Z][0-9]{3}$/", $c->getCityCode()));
                } catch (\Exception $ex) {
                    echo $row['codiceFiscale'] . " - " . $ex->getMessage() . "\n";
                }
            }
        }
    }
}


readme_examples/examples.php
<?php
include __DIR__.'/../vendor/autoload.php';


use IvanoMatteo\CodiceFiscale\CodiceFiscale;



try{

    // will raise an exception if the format is not valid
    $c = CodiceFiscale::parse("RSSMRAULRL1H50MM",1900); // passing the "century" arg, help to validate in case of leap years
    $c = CodiceFiscale::parse("RSSMRAULRL1H50MM");

    echo "\n"."matchName ".($c->matchName('Mario')? 'si' : 'no');
    echo "\n"."isOmocodia ".($c->isOmocodia()? 'si' : 'no');

    // to extract the date of birth, the "century" argumenti is required
    echo "\n"."getDateOfBirth ".$c->getDateOfBirth(1900)->format('Y-m-d');

    // is possible to extract the raw date of birth:  { "year":"yy", "month":"mm", "day":"dd" }
    echo "\n"."getDateOfBirthObj ".json_encode($c->getDateOfBirthRaw());

    // the library can provide the nearest date of birth according to current date:
    echo "\n"."getProbableDateOfBirth ".$c->getProbableDateOfBirth()->format('Y-m-d');
    // also according to minimum age and current date arguments
    echo "\n"."getProbableDateOfBirth 18 ".$c->getProbableDateOfBirth(18,'2019-01-01')->format('Y-m-d');

}catch(\Exception $ex){
  echo $ex->getMessage();
}


//
// calculate using person data 
//
$name = 'Mario';
$familyName = 'Rossi';
$dateOfBirth = '1980-10-01';
$sex = 'M';
$cityCode = 'H501';

$cf = CodiceFiscale::calculate($name, $familyName, $dateOfBirth, $sex, $cityCode);

//
// calcumate using an array (it accept an object as well)
//
$person = [
    'name' => $name,
    'familyName' => $familyName,
    'date' => $dateOfBirth,
    'sex' => $sex,
    'cityCode' => $cityCode,
];
$map = [  // it possible to provide a map, to match fields with different names
    'dateOfBirth' => 'date'
];
$cfx = CodiceFiscale::calculateObj($person, $map);


// generate all the 127 possible "omocodia" variations
$variazioni = $cf->generateVariations();

// generate variation n. 7 only
$variazioni = $cf->generateVariations(7);


